ipinfo/cmd_download.go

```diff
@@ -0,0 +1,225 @@
+		"--format":   predict.Nothing,
+		"-t":         predict.Nothing,
+		"--token":    predict.Nothing,
+		"-h":         predict.Nothing,
+		"--help":     predict.Nothing,
+	},
+}
+
+func printHelpDownload() {
+	fmt.Printf(
+		`Usage: %s download [<database>] [<opts>]
+
+Description:
+    Download the free ipinfo databases.
+
+Examples:
+    # Download country database in csv format.
+    $ %[1]s download country -f csv > country.csv
+    $ %[1]s download country-asn country_asn.mmdb
+
+Databases:
+    asn            free ipinfo asn database.
+    country        free ipinfo country database.
+    country-asn    free ipinfo country-asn database.
+
+Options:
+  General:
+    --token <tok>, -t <tok>
+      use <tok> as API token.
+    --help, -h
+      show help.
+
+Outputs:
+	--compress, -c
+	save the file in compressed format.
+	default: false.
+    --format , -f <mmdb | json | csv>
+    output format of the database file.
+    default: mmdb.
+`, progBase)
+}
+
+func cmdDownload() error {
+	var fTok string
+	var fFmt string
+	var fZip bool
+	var fHelp bool
+
+	pflag.StringVarP(&fTok, "token", "t", "", "the token to use.")
+	pflag.StringVarP(&fFmt, "format", "f", "mmdb", "the output format to use.")
+	pflag.BoolVarP(&fZip, "compress", "c", false, "compressed output.")
+	pflag.BoolVarP(&fHelp, "help", "h", false, "show help.")
+	pflag.Parse()
+
+	args := pflag.Args()[1:]
+	if fHelp || len(args) > 2 || len(args) < 1 {
+		printHelpDownload()
+		return nil
+	}
+
+	var token string
+	if fTok == "" {
+		token = gConfig.Token
+	}
+
+	// require token for download.
+	if token == "" {
+		return errors.New("downloading requires a token; login via `ipinfo login` or pass the `--token` argument")
+	}
+
+	// get download format.
+	var format string
+	switch strings.ToLower(fFmt) {
+	case "mmdb":
+		format = "mmdb"
+	case "csv":
+		format = "csv.gz"
+	case "json":
+		format = "json.gz"
+	default:
+		return errors.New("unknown download format")
+	}
+
+	// download the db.
+	switch strings.ToLower(args[0]) {
+	case "asn":
+		err := downloadDb("asn", format, token, fZip)
+		if err != nil {
+			return err
+		}
+	case "country":
+		err := downloadDb("country", format, token, fZip)
+		if err != nil {
+			return err
+		}
+	case "country-asn":
+		err := downloadDb("country_asn", format, token, fZip)
+		if err != nil {
+			return err
+		}
+	default:
+		return fmt.Errorf("database '%v' is invalid", args[0])
+	}
+
+	return nil
+}
+
+func downloadDb(name, format, token string, zip bool) error {
+	url := fmt.Sprintf("%s%s.%s?token=%s", dbDownloadURL, name, format, token)
+
+	// make API req to download the file.
+	res, err := http.Get(url)
+	if err != nil {
+		return err
+	}
+	defer res.Body.Close()
+
+	// if output not terminal unzip and write to stdout.
+	if fileInfo, _ := os.Stdout.Stat(); (fileInfo.Mode() & os.ModeCharDevice) == 0 {
+		err := unzipWrite(os.Stdout, res.Body)
+		if err != nil {
+			return err
+		}
+	} else {
+		// get filename.
```

Comment:
Let us try to get this filename earlier in the code. We do this in eg. `mmdbctl import` as it greatly simplifies the logic  flow: do all the flag/etc stuff upfront as much as possible, do basic checks on them, initiate all the clients needed, begin business logic.